File: go/test/team_entity_test.go
package sdktest

import (
	"encoding/json"
	"fmt"
	"os"
	"path/filepath"
	"runtime"
	"strings"
	"testing"
	"time"

	sdk "github.com/voxgig-sdk/github-sdk/go"
	"github.com/voxgig-sdk/github-sdk/go/core"

	vs "github.com/voxgig-sdk/github-sdk/go/utility/struct"
)

func TestTeamEntity(t *testing.T) {
	t.Run("instance", func(t *testing.T) {
		testsdk := sdk.TestSDK(nil, nil)
		ent := testsdk.Team(nil)
		if ent == nil {
			t.Fatal("expected non-nil TeamEntity")
		}
	})

	// Feature #4: the entity Stream(action, ...) method runs the op pipeline and
	// returns a channel over result items. With the streaming feature active it
	// yields the feature's incremental output; otherwise it falls back to the
	// materialised list so Stream always yields.
	t.Run("stream", func(t *testing.T) {
		seed := map[string]any{
			"entity": map[string]any{
				"team": map[string]any{
					"s1": map[string]any{"id": "s1"},
					"s2": map[string]any{"id": "s2"},
					"s3": map[string]any{"id": "s3"},
				},
			},
		}

		// Fallback: streaming inactive -> yields the materialised list items.
		base := sdk.TestSDK(seed, nil)
		var seen []any
		for item := range base.Team(nil).Stream("list", nil, nil) {
			seen = append(seen, item)
		}
		if len(seen) != 3 {
			t.Fatalf("expected 3 streamed items, got %d", len(seen))
		}

		// Inbound: streaming active -> yields each item from the feature iterator.
		hasStreaming := false
		if fm, ok := core.SharedConfig()["feature"].(map[string]any); ok {
			_, hasStreaming = fm["streaming"]
		}
		if hasStreaming {
			streamSdk := sdk.TestSDK(seed, map[string]any{
				"feature": map[string]any{"streaming": map[string]any{"active": true}},
			})
			var got []any
			for item := range streamSdk.Team(nil).Stream("list", nil, nil) {
				if sub, ok := item.([]any); ok {
					got = append(got, sub...)
				} else {
					got = append(got, item)
				}
			}
			if len(got) != 3 {
				t.Fatalf("expected 3 items via streaming feature, got %d", len(got))
			}
		}
	})

	t.Run("basic", func(t *testing.T) {
		setup := teamBasicSetup(nil)
		// Per-op sdk-test-control.json skip — basic test exercises a flow
		// with multiple ops; skipping any op skips the whole flow.
		_mode := "unit"
		if setup.live {
			_mode = "live"
		}
		for _, _op := range []string{"create", "list", "update", "load", "remove"} {
			if _shouldSkip, _reason := isControlSkipped("entityOp", "team." + _op, _mode); _shouldSkip {
				if _reason == "" {
					_reason = "skipped via sdk-test-control.json"
				}
				t.Skip(_reason)
				return
			}
		}
		// The basic flow consumes synthetic IDs from the fixture. In live mode
		// without an *_ENTID env override, those IDs hit the live API and 4xx.
		if setup.syntheticOnly {
			t.Skip("live entity test uses synthetic IDs from fixture — set GITHUB_TEST_TEAM_ENTID JSON to run live")
			return
		}
		client := setup.client

		// CREATE
		teamRef01Ent := client.Team(nil)
		teamRef01Data := core.ToMapAny(vs.GetProp(
			vs.GetPath(setup.data, []any{"new", "team"}), "team_ref01"))
		teamRef01Data["branch_id"] = setup.idmap["branch01"]
		teamRef01Data["discussion_id"] = setup.idmap["discussion01"]
		teamRef01Data["discussion_number"] = setup.idmap["discussion_number01"]
		teamRef01Data["invitation_id"] = setup.idmap["invitation01"]
		teamRef01Data["org"] = setup.idmap["org01"]
		teamRef01Data["org_id"] = setup.idmap["org01"]
		teamRef01Data["organization_role_id"] = setup.idmap["organization_role01"]
		teamRef01Data["owner"] = setup.idmap["owner01"]
		teamRef01Data["repo"] = setup.idmap["repo01"]
		teamRef01Data["team_id"] = setup.idmap["team01"]
		teamRef01Data["team_slug"] = setup.idmap["team_slug01"]

		teamRef01DataResult, err := teamRef01Ent.Create(teamRef01Data, nil)
		if err != nil {
			t.Fatalf("create failed: %v", err)
		}
		teamRef01Data = core.ToMapAny(entityData(teamRef01DataResult))
		if teamRef01Data == nil {
			t.Fatal("expected create result to be a map")
		}
		if teamRef01Data["id"] == nil {
			t.Fatal("expected created entity to have an id")
		}

		// LIST
		teamRef01Match := map[string]any{}

		teamRef01ListResult, err := teamRef01Ent.List(teamRef01Match, nil)
		if err != nil {
			t.Fatalf("list failed: %v", err)
		}
		teamRef01List, teamRef01ListOk := teamRef01ListResult.([]any)
		if !teamRef01ListOk {
			t.Fatalf("expected list result to be an array, got %T", teamRef01ListResult)
		}

		foundItem := vs.Select(entityListToData(teamRef01List), map[string]any{"id": teamRef01Data["id"]})
		if vs.IsEmpty(foundItem) {
			t.Fatal("expected to find created entity in list")
		}

		// UPDATE
		teamRef01DataUp0Up := map[string]any{
			"id": teamRef01Data["id"],
		}

		teamRef01MarkdefUp0Name := "archive_url"
		teamRef01MarkdefUp0Value := fmt.Sprintf("Mark01-team_ref01_%d", setup.now)
		teamRef01DataUp0Up[teamRef01MarkdefUp0Name] = teamRef01MarkdefUp0Value

		teamRef01ResdataUp0Result, err := teamRef01Ent.Update(teamRef01DataUp0Up, nil)
		if err != nil {
			t.Fatalf("update failed: %v", err)
		}
		teamRef01ResdataUp0 := core.ToMapAny(entityData(teamRef01ResdataUp0Result))
		if teamRef01ResdataUp0 == nil {
			t.Fatal("expected update result to be a map")
		}
		if teamRef01ResdataUp0["id"] != teamRef01DataUp0Up["id"] {
			t.Fatal("expected update result id to match")
		}
		if teamRef01ResdataUp0[teamRef01MarkdefUp0Name] != teamRef01MarkdefUp0Value {
			t.Fatalf("expected %s to be updated, got %v", teamRef01MarkdefUp0Name, teamRef01ResdataUp0[teamRef01MarkdefUp0Name])
		}

		// LOAD
		teamRef01MatchDt0 := map[string]any{
			"id": teamRef01Data["id"],
		}
		teamRef01DataDt0Loaded, err := teamRef01Ent.Load(teamRef01MatchDt0, nil)
		if err != nil {
			t.Fatalf("load failed: %v", err)
		}
		teamRef01DataDt0LoadResult := core.ToMapAny(entityData(teamRef01DataDt0Loaded))
		if teamRef01DataDt0LoadResult == nil {
			t.Fatal("expected load result to be a map")
		}
		if teamRef01DataDt0LoadResult["id"] != teamRef01Data["id"] {
			t.Fatal("expected load result id to match")
		}

		// REMOVE
		teamRef01MatchRm0 := map[string]any{
			"id": teamRef01Data["id"],
		}
		_, err = teamRef01Ent.Remove(teamRef01MatchRm0, nil)
		if err != nil {
			t.Fatalf("remove failed: %v", err)
		}

		// LIST
		teamRef01MatchRt0 := map[string]any{}

		teamRef01ListRt0Result, err := teamRef01Ent.List(teamRef01MatchRt0, nil)
		if err != nil {
			t.Fatalf("list failed: %v", err)
		}
		teamRef01ListRt0, teamRef01ListRt0Ok := teamRef01ListRt0Result.([]any)
		if !teamRef01ListRt0Ok {
			t.Fatalf("expected list result to be an array, got %T", teamRef01ListRt0Result)
		}

		notFoundItem := vs.Select(entityListToData(teamRef01ListRt0), map[string]any{"id": teamRef01Data["id"]})
		if !vs.IsEmpty(notFoundItem) {
			t.Fatal("expected removed entity to not be in list")
		}

	})
}

func teamBasicSetup(extra map[string]any) *entityTestSetup {
	loadEnvLocal()

	_, filename, _, _ := runtime.Caller(0)
	dir := filepath.Dir(filename)

	entityDataFile := filepath.Join(dir, "..", "..", ".sdk", "test", "entity", "team", "TeamTestData.json")

	entityDataSource, err := os.ReadFile(entityDataFile)
	if err != nil {
		panic("failed to read team test data: " + err.Error())
	}

	var entityData map[string]any
	if err := json.Unmarshal(entityDataSource, &entityData); err != nil {
		panic("failed to parse team test data: " + err.Error())
	}

	options := map[string]any{}
	options["entity"] = entityData["existing"]

	client := sdk.TestSDK(options, extra)

	// Generate idmap via transform, matching TS pattern.
	idmap, _ := vs.Transform(
		[]any{"team01", "team02", "team03", "org01", "org02", "org03", "member01", "member02", "member03", "invitation01", "invitation02", "invitation03", "organization_role01", "organization_role02", "organization_role03", "repo01", "repo02", "repo03", "branch01", "branch02", "branch03", "discussion01", "discussion02", "discussion03", "membership01", "membership02", "membership03", "project01", "project02", "project03", "comment01", "comment02", "comment03", "discussion_number01", "owner01", "team_slug01"},
		map[string]any{
			"`$PACK`": []any{"", map[string]any{
				"`$KEY`": "`$COPY`",
				"`$VAL`": []any{"`$FORMAT`", "upper", "`$COPY`"},
			}},
		},
	)

	// Detect ENTID env override before envOverride consumes it. When live
	// mode is on without a real override, the basic test runs against synthetic
	// IDs from the fixture and 4xx's. Surface this so the test can skip.
	entidEnvRaw := os.Getenv("GITHUB_TEST_TEAM_ENTID")
	idmapOverridden := entidEnvRaw != "" && strings.HasPrefix(strings.TrimSpace(entidEnvRaw), "{")

	env := envOverride(map[string]any{
		"GITHUB_TEST_TEAM_ENTID": idmap,
		"GITHUB_TEST_LIVE":      "FALSE",
		"GITHUB_TEST_EXPLAIN":   "FALSE",
		"GITHUB_APIKEY":         "",
	})

	idmapResolved := core.ToMapAny(env["GITHUB_TEST_TEAM_ENTID"])
	if idmapResolved == nil {
		idmapResolved = core.ToMapAny(idmap)
	}

	if env["GITHUB_TEST_LIVE"] == "TRUE" {
		// An empty map, not a nil one: Merge returns nil when its last entry
		// is nil, and BasicSetup is normally called with no extras - so a
		// bare nil silently discarded the apikey and server values below.
		extraOpts := extra
		if extraOpts == nil {
			extraOpts = map[string]any{}
		}

		mergedOpts := vs.Merge([]any{
			// liveClientOptions() FIRST, so the generated fields below win:
			// sdk-test-control.json's test.client.options adds to the live
			// client, it does not redirect it.
			liveClientOptions(),
			map[string]any{
				"apikey": env["GITHUB_APIKEY"],
			},
			extraOpts,
		})
		client = sdk.NewGithubSDK(core.ToMapAny(mergedOpts))
	}

	live := env["GITHUB_TEST_LIVE"] == "TRUE"
	return &entityTestSetup{
		client:        client,
		data:          entityData,
		idmap:         idmapResolved,
		env:           env,
		explain:       env["GITHUB_TEST_EXPLAIN"] == "TRUE",
		live:          live,
		syntheticOnly: live && !idmapOverridden,
		now:           time.Now().UnixMilli(),
	}
}

File: ts/test/entity/team/TeamEntity.test.ts


import Path from 'node:path'
import * as Fs from 'node:fs'

import { test, describe, afterEach } from 'node:test'
import assert from 'node:assert'


import { GithubSDK, BaseFeature, stdutil } from '../../..'

import {
  envOverride,
  liveClientOptions,
  liveDelay,
  loadEnvLocal,
  makeCtrl,
  makeMatch,
  makeReqdata,
  makeStepData,
  makeValid,
  maybeSkipControl,
} from '../../utility'


// AFTER the imports on purpose: TypeScript hoists `import` above any
// statement in the emitted CommonJS, so a loader placed above them would
// run only after every imported module had already been evaluated - and
// anything reading process.env at module scope would miss these values.
loadEnvLocal(__dirname + '/../../../.env.local')


describe('TeamEntity', async () => {

  // Per-test live pacing. Delay is read from sdk-test-control.json's
  // `test.live.delayMs`; only sleeps when GITHUB_TEST_LIVE=TRUE.
  afterEach(liveDelay('GITHUB_TEST_LIVE'))

  test('instance', async () => {
    const testsdk = GithubSDK.test()
    const ent = testsdk.Team()
    assert(null != ent)
  })


  test('basic', async (t) => {

    const live = 'TRUE' === process.env.GITHUB_TEST_LIVE
    for (const op of ['create', 'list', 'update', 'load', 'remove']) {
      if (maybeSkipControl(t, 'entityOp', 'team.' + op, live)) return
    }

    const setup = basicSetup()
    // The basic flow consumes synthetic IDs and field values from the
    // fixture (entity TestData.json). Those don't exist on the live API.
    // Skip live runs unless the user provided a real ENTID env override.
    if (setup.syntheticOnly) {
      t.skip('live entity test uses synthetic IDs from fixture — set GITHUB_TEST_TEAM_ENTID JSON to run live')
      return
    }
    const client = setup.client
    const struct = setup.struct

    const isempty = struct.isempty
    const select = struct.select


    // CREATE
    const team_ref01_ent = client.Team()
    let team_ref01_data = setup.data.new.team['team_ref01']
    team_ref01_data['branch_id'] = setup.idmap['branch01']
    team_ref01_data['discussion_id'] = setup.idmap['discussion01']
    team_ref01_data['discussion_number'] = setup.idmap['discussion_number01']
    team_ref01_data['invitation_id'] = setup.idmap['invitation01']
    team_ref01_data['org'] = setup.idmap['org01']
    team_ref01_data['org_id'] = setup.idmap['org01']
    team_ref01_data['organization_role_id'] = setup.idmap['organization_role01']
    team_ref01_data['owner'] = setup.idmap['owner01']
    team_ref01_data['repo'] = setup.idmap['repo01']
    team_ref01_data['team_id'] = setup.idmap['team01']
    team_ref01_data['team_slug'] = setup.idmap['team_slug01']

    team_ref01_data = (await team_ref01_ent.create(team_ref01_data)).data()
    assert(null != team_ref01_data.id)


    // LIST
    const team_ref01_match: any = {}

    const team_ref01_list = (await team_ref01_ent.list(team_ref01_match)).map((e: any) => e.data())

    assert(!isempty(select(team_ref01_list, { id: team_ref01_data.id })))


    // UPDATE
    const team_ref01_data_up0: any = {}
    team_ref01_data_up0.id = team_ref01_data.id

    const team_ref01_markdef_up0 = { name: 'archive_url', value: 'Mark01-team_ref01_' + setup.now }
    ;(team_ref01_data_up0 as any)[team_ref01_markdef_up0.name] = team_ref01_markdef_up0.value

    const team_ref01_resdata_up0 = (await team_ref01_ent.update(team_ref01_data_up0)).data()
    assert(team_ref01_resdata_up0.id === team_ref01_data_up0.id)

    assert((team_ref01_resdata_up0 as any)[team_ref01_markdef_up0.name] === team_ref01_markdef_up0.value)


    // LOAD
    const team_ref01_match_dt0: any = {}
    team_ref01_match_dt0.id = team_ref01_data.id
    const team_ref01_data_dt0 = (await team_ref01_ent.load(team_ref01_match_dt0)).data()
    assert(team_ref01_data_dt0.id === team_ref01_data.id)


    // REMOVE
    const team_ref01_match_rm0: any = { id: team_ref01_data.id }
    await team_ref01_ent.remove(team_ref01_match_rm0)
  

    // LIST
    const team_ref01_match_rt0: any = {}

    const team_ref01_list_rt0 = (await team_ref01_ent.list(team_ref01_match_rt0)).map((e: any) => e.data())

    assert(isempty(select(team_ref01_list_rt0, { id: team_ref01_data.id })))


  })
})



function basicSetup(extra?: any) {
  // TODO: fix test def options
  const options: any = {} // null

  // TODO: needs test utility to resolve path
  const entityDataFile =
    Path.resolve(__dirname, 
      '../../../../.sdk/test/entity/team/TeamTestData.json')

  // TODO: file ready util needed?
  const entityDataSource = Fs.readFileSync(entityDataFile).toString('utf8')

  // TODO: need a xlang JSON parse utility in voxgig/struct with better error msgs
  const entityData = JSON.parse(entityDataSource)

  options.entity = entityData.existing

  let client = GithubSDK.test(options, extra)
  const struct = client.utility().struct
  const merge = struct.merge
  const transform = struct.transform

  let idmap = transform(
    ['team01','team02','team03','org01','org02','org03','member01','member02','member03','org01','org02','org03','invitation01','invitation02','invitation03','org01','org02','org03','organization_role01','organization_role02','organization_role03','org01','org02','org03','team01','team02','team03','repo01','repo02','repo03','branch01','branch02','branch03','org01','org02','org03','team01','team02','team03','discussion01','discussion02','discussion03','org01','org02','org03','team01','team02','team03','membership01','membership02','membership03','org01','org02','org03','team01','team02','team03','project01','project02','project03','org01','org02','org03','team01','team02','team03','repo01','repo02','repo03','org01','org02','org03','team01','team02','team03','discussion01','discussion02','discussion03','comment01','comment02','comment03'],
    {
      '`$PACK`': ['', {
        '`$KEY`': '`$COPY`',
        '`$VAL`': ['`$FORMAT`', 'upper', '`$COPY`']
      }]
    })

  // Detect whether the user provided a real ENTID JSON via env var. The
  // basic flow consumes synthetic IDs from the fixture file; without an
  // override those synthetic IDs reach the live API and 4xx. Surface this
  // to the test so it can skip rather than fail.
  const idmapEnvVal = process.env['GITHUB_TEST_TEAM_ENTID']
  const idmapOverridden = null != idmapEnvVal && idmapEnvVal.trim().startsWith('{')

  const env = envOverride({
    'GITHUB_TEST_TEAM_ENTID': idmap,
    'GITHUB_TEST_LIVE': 'FALSE',
    'GITHUB_TEST_EXPLAIN': 'FALSE',
  })

  idmap = env['GITHUB_TEST_TEAM_ENTID']

  const live = 'TRUE' === env.GITHUB_TEST_LIVE

  if (live) {
    client = new GithubSDK(merge([
      // FIRST, so the generated fields below win: sdk-test-control.json's
      // test.client.options adds to the live client, it does not redirect it.
      liveClientOptions(),
      {
      },
      // 'extra || {}', not a bare 'extra': struct.merge returns UNDEFINED when the
      // last entry is undefined, and basicSetup is normally called with no
      // argument at all - so a bare 'extra' silently discarded the apikey
      // and server values above and handed the SDK undefined. Harmless
      // while there was nothing in that object; not harmless now.
      extra || {}
    ]))
  }

  const setup = {
    idmap,
    env,
    options,
    client,
    struct,
    data: entityData,
    explain: 'TRUE' === env.GITHUB_TEST_EXPLAIN,
    live,
    syntheticOnly: live && !idmapOverridden,
    now: Date.now(),
  }

  return setup
}
  
